engineering-decision-records/011-deterministic-simulation-testing.md

```diff
@@ -0,0 +1,279 @@
+  nodes, so implementing crashes and restarts can be done by dropping and
+  respawning a node;
+
+* I/O and GC pauses. Rust doesn't have GC, but disk I/O can sometimes be slower
+  than normal due to other stuff happening on the computer that the node is
+  running on. I/O pauses can be implemented at the pure-stage level by delaying
+  when a future becomes available;
+
+* Time skews (the clock of nodes drifts a part). The simulator controls the
+  clock of the nodes, via pure-stage's API, so skews can be implemented by
+  advancing some node's clocks more than others;
+
+* Disk failures (since all of our stable storage is delegated to RocksDB we
+  inherit its storage failure model). To test this the simulator can inject the
+  following disk faults on nodes:
+    - Delete the whole database;
+    - Flip bits in the database;
+    - Truncate the write-ahead-log of the database (to simulate a crash where
+      the last entry wasn't fully fsynced to disk). For more details, see
+      [Protocol-Aware Recovery for Consensus-Based Distributed
+      Storage](https://dl.acm.org/doi/10.1145/3241062).
+
+#### Byzantine faults
+
+The space of possibilities here is infinite. Any combination of changing any
+message or any piece of state on a faulty node is allowed. Although the
+protocol, with its use of cryptographically secure hashes, should render all of
+these attacks useless, nevertheless this needs to be tested to some extent.
+
+As a first approximation one could have the simulator mutate messages between
+or the stable storage of a node to ensure that hashes are checked.
+
+Coordinated attacks that explore the extremes of protocol parameters and timings
+of messages is out of scope of this decision. Let us just note there's the
+possibility to add Byzantine (malicious) nodes to the simulation test and
+otherwise run the tests as before, and ensure that the non-faulty nodes still
+work as intended.
+
+It should also be noted that this space contains admin mistakes, such as
+misconfigurations of node, computer or network settings. We've not thought
+about how to simulate these type of faults yet.
+
+### Membership changes and upgrades
+
+Testing how the system behaves as nodes leave and join, potentially with a
+different version of the software, is also important as there could be
+backward- and forward-compatibility problems between the version.
+
+While we've not put much thought into this yet, the simulator has been designed
+in such a way that it can communicate with the nodes via stdin/stdout a la
+Jepsen's Maelstrom, which means we could technically run different versions of
+the node binaries (this would be slower than the default in-process tests).
+
+### Test case generation
+
+Due to the consensus protocol's requirement that all messages need to be hashed
+and refer to the previous message's hash (i.e. hash-chained), it's not possible
+to generate "random" traffic as one normally would with property-based testing.
```

Comment:
It sure isn’t straight-forward and the libraries aren’t made for this, meaning it will be inconvenient, but the “not possible” part of the statement is incorrect. It is a matter of generating test inputs that then are used to compute a resulting block tree with the correct hashes, e.g. using an integer to point to a previous block (using modulo arithmetic to adequately “shorten” it).